vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go

```diff
@@ -0,0 +1,162 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	k8sapiv1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/model"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	externalmetricsv1beta1 "k8s.io/metrics/pkg/apis/external_metrics/v1beta1"
+	"k8s.io/metrics/pkg/apis/metrics/v1beta1"
+	"time"
+
+	resourceclient "k8s.io/metrics/pkg/client/clientset/versioned/typed/metrics/v1beta1"
+	"k8s.io/metrics/pkg/client/external_metrics"
+)
+
+// PodMetricsLister wraps both metrics-client and External Metrics
+type PodMetricsLister interface {
+	List(ctx context.Context, namespace string, model *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error)
+}
+
+type podMetricsSource struct {
+	metricsGetter resourceclient.PodMetricsesGetter
+}
+
+// NewPodMetricsesSource Returns a Source-wrapper around PodMetricsesGetter.
+func NewPodMetricsesSource(source resourceclient.PodMetricsesGetter) PodMetricsLister {
+	return podMetricsSource{metricsGetter: source}
+}
+
+func (s podMetricsSource) List(ctx context.Context, namespace string, _ *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	podMetricsInterface := s.metricsGetter.PodMetricses(namespace)
+	return podMetricsInterface.List(ctx, opts)
+}
+
+type externalMetricsClient struct {
+	externalClient external_metrics.ExternalMetricsClient
+	options        ExternalClientOptions
+	clusterState   *model.ClusterState
+}
+
+// ExternalClientOptions specifies parameters for using an External Metrics Client.
+type ExternalClientOptions struct {
+	ResourceMetrics                                  map[k8sapiv1.ResourceName]string
+	PodNamespaceLabel, PodNameLabel                  string
+	CtrNamespaceLabel, CtrPodNameLabel, CtrNameLabel string
+}
+
+// NewExternalClient returns a Source for an External Metrics Client.
+func NewExternalClient(c *rest.Config, clusterState *model.ClusterState, options ExternalClientOptions) PodMetricsLister {
+	extClient, err := external_metrics.NewForConfig(c)
+	if err != nil {
+		klog.Fatalf("Failed initializing external metrics client: %v", err)
+	}
+	return &externalMetricsClient{
+		externalClient: extClient,
+		options:        options,
+		clusterState:   clusterState,
+	}
+}
+
+type podContainerResourceMap map[model.PodID]map[string]k8sapiv1.ResourceList
+
+func (s *externalMetricsClient) addMetrics(list *externalmetricsv1beta1.ExternalMetricValueList, name k8sapiv1.ResourceName, resourceMap podContainerResourceMap, namespace, podName string) {
```

Comment:
I'd change this to

```
func (resourceMap *podContainerResourceMap) addMetrics(list *externalmetricsv1beta1.ExternalMetricValueList, name k8sapiv1.ResourceName, namespace, podName ctrNameLabel, string)
```

since it's adding stuff to podContainerResourceMap. The way it's now I thought it was adding things to `externalMetricsClient` and was very confused why we'd do that

Reply:
Changed.